cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +217,246 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
   CUDA_TRY(cub::DeviceSegmentedReduce::Sum(nullptr,
                                            temp_storage_bytes,
-                                           word_num_set_bits,
-                                           d_null_counts.begin(),
+                                           num_set_bits_in_word,
+                                           d_bit_counts.begin(),
                                            num_ranges,
                                            first_word_indices,
                                            last_word_indices,
                                            stream.value()));
   rmm::device_buffer d_temp_storage(temp_storage_bytes, stream);
 
-  // second perform segmented reduction
-
+  // Perform segmented reduction.
   CUDA_TRY(cub::DeviceSegmentedReduce::Sum(d_temp_storage.data(),
                                            temp_storage_bytes,
-                                           word_num_set_bits,
-                                           d_null_counts.begin(),
+                                           num_set_bits_in_word,
+                                           d_bit_counts.begin(),
                                            num_ranges,
                                            first_word_indices,
                                            last_word_indices,
                                            stream.value()));
 
   CHECK_CUDA(stream.value());
 
-  // third, adjust counts in segment boundaries (if segments are not
-  // word-aligned)
-
+  // Adjust counts in segment boundaries (if segments are not word-aligned).
   constexpr size_type block_size{256};
-
   cudf::detail::grid_1d grid(num_ranges, block_size);
-
   subtract_set_bits_range_boundaries_kernel<<<grid.num_blocks,
                                               grid.num_threads_per_block,
                                               0,
                                               stream.value()>>>(
-    bitmask, num_ranges, d_first_indices.begin(), d_last_indices.begin(), d_null_counts.begin());
+    bitmask, num_ranges, first_bit_indices_begin, last_bit_indices_begin, d_bit_counts.begin());
+
+  if (count_bits == count_bits_policy::UNSET_BITS) {
+    // Convert from set bits counts to unset bits by subtracting the number of
+    // set bits from the length of the segment.
+    thrust::for_each(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator(0),
+                     thrust::make_counting_iterator(static_cast<size_type>(d_bit_counts.size())),
+                     [first_bit_indices_begin,
+                      last_bit_indices_begin,
+                      d_bit_counts = d_bit_counts.data()] __device__(size_type i) {
+                       auto const begin = *(first_bit_indices_begin + i);
+                       auto const end   = *(last_bit_indices_begin + i);
+                       d_bit_counts[i]  = (end - begin) - d_bit_counts[i];
+                     });
```

Comment:
```suggestion
    auto segments_begin = thrust::zip_iterator(thrust::make_tuple(first_bit_indices_begin, last_bit_indices_begin));
    auto segments_size = thrust::transform_iterator(segments_begin, [] __device__ (auto segment){
       auto const [begin,end] = segment;
       return end - begin;
    });
    thrust::transform(rmm::exec_policy(stream),
                     segments_size, 
                     segments_size + num_ranges,
                     d_bit_counts.begin(),
                     [] __device__(auto segment_size, auto segment_bit_count) {
                       return segment_size - segment_bit_count;
                     });
```

Reply:
I applied this change (with a couple minor fixes) in 65e3745dfc6fc52c3e28ea7d5e13bd7dff9b5430. However, I wasn't able to make the structured binding compile. @jrhemstad if you have any insight on this, feel free to take a peek.

Reply:
Here's a minimal reproduction of the issue with structured bindings. It seems like this might be a bug in thrust (or simply that structured bindings aren't yet implemented for `thrust::tuple`?). https://godbolt.org/z/rvYoPb948

I also tested locally with Thrust 1.15.0 and got the same problem.

Reply:
This issue was also reported internally and it should be resolved by https://github.com/NVIDIA/thrust/issues/1212.